File: dsa-java/13-linked-list/easy/09_IntersectionOfTwoLinkedLists.java
/*
 * Problem Statement:
 * Given the heads of two singly linked-lists headA and headB, return the node at which
 * the two lists intersect. If the two linked lists have no intersection at all, return null.
 * Note that the linked lists must retain their original structure after the function returns.
 * Could you write a solution that runs in O(m + n) time and use only O(1) memory?
 * (LeetCode 160: Intersection of Two Linked Lists)
 * 
 * Asked in: Amazon, Microsoft, Bloomberg, Apple, Facebook/Meta
 */

class IntersectionOfTwoLinkedLists {

    static class ListNode {
        int val;
        ListNode next;
        ListNode(int val) { this.val = val; }
    }

    /**
     * Approach:
     * Two Pointers Wrapping Around:
     * Pointer pA traverses listA then wraps to headB.
     * Pointer pB traverses listB then wraps to headA.
     * Since len(A) + len(B) == len(B) + len(A), both pointers traverse the exact same total length.
     * They will either collide at the intersection node or both hit null simultaneously!
     */
    public static ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) return null;

        ListNode pA = headA;
        ListNode pB = headB;

        while (pA != pB) {
            pA = (pA == null) ? headB : pA.next;
            pB = (pB == null) ? headA : pB.next;
        }

        return pA; // Intersection node or null
    }

    public static void main(String[] args) {
        // Shared tail: 8 -> 4 -> 5
        ListNode common = new ListNode(8);
        common.next = new ListNode(4);
        common.next.next = new ListNode(5);

        // List A: 4 -> 1 -> common
        ListNode headA = new ListNode(4);
        headA.next = new ListNode(1);
        headA.next.next = common;

        // List B: 5 -> 6 -> 1 -> common
        ListNode headB = new ListNode(5);
        headB.next = new ListNode(6);
        headB.next.next = new ListNode(1);
        headB.next.next.next = common;

        ListNode intersection = getIntersectionNode(headA, headB);
        System.out.println("Intersection node value: " + (intersection != null ? intersection.val : "null") + " (Expected: 8)");
    }
}

/*
 * Time Complexity: O(M + N) where M and N are lengths of lists A and B.
 * Space Complexity: O(1) auxiliary pointers.
 */

File: dsa-c/13-linked-list/easy/09_IntersectionOfTwoLinkedLists.c
/*
 * Problem Statement:
 * Given the heads of two singly linked-lists headA and headB, return the node at which
 * the two lists intersect. If the two linked lists have no intersection at all, return null.
 * Note that the linked lists must retain their original structure after the function returns.
 * Could you write a solution that runs in O(m + n) time and use only O(1) memory?
 * (LeetCode 160: Intersection of Two Linked Lists)
 * 
 * Asked in: Amazon, Microsoft, Bloomberg, Apple, Facebook/Meta
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <limits.h>
#include <math.h>
#include <stdint.h>
#include <float.h>
#include <ctype.h>

#ifndef MAX
#define MAX(a, b) (((a) > (b)) ? (a) : (b))
#endif

#ifndef MIN
#define MIN(a, b) (((a) < (b)) ? (a) : (b))
#endif

static const char* toBinaryString(int n) {
    static char buf[33];
    buf[32] = '\0';
    for (int i = 31; i >= 0; i--) {
        buf[31 - i] = ((n >> i) & 1) ? '1' : '0';
    }
    return buf;
}

typedef struct ListNode {
    int val;
    struct ListNode* next;
} ListNode;

static ListNode* createNode(int val) {
    ListNode* node = (ListNode*)malloc(sizeof(ListNode));
    if (!node) exit(1);
    node->val = val;
    node->next = NULL;
    return node;
}

static void freeList(ListNode* head) {
    while (head) {
        ListNode* tmp = head;
        head = head->next;
        free(tmp);
    }
}

static ListNode* buildList(int count, int vals[]) {
    if (count <= 0) return NULL;
    ListNode* head = createNode(vals[0]);
    ListNode* curr = head;
    for (int i = 1; i < count; i++) {
        curr->next = createNode(vals[i]);
        curr = curr->next;
    }
    return head;
}

/**
     * Approach:
     * Two Pointers Wrapping Around:
     * Pointer pA traverses listA then wraps to headB.
     * Pointer pB traverses listB then wraps to headA.
     * Since len(A) + len(B) == len(B) + len(A), both pointers traverse the exact same total length.
     * They will either collide at the intersection node or both hit NULL simultaneously!
     */
    ListNode* getIntersectionNode(ListNode* headA, ListNode* headB) {
        if (headA == NULL || headB == NULL) return NULL;

        ListNode* pA = headA;
        ListNode* pB = headB;

        while (pA != pB) {
            pA = (pA == NULL) ? headB : pA->next;
            pB = (pB == NULL) ? headA : pB->next;
        }

        return pA; // Intersection node or NULL
    }

    int main(void) {
        // Shared tail: 8 -> 4 -> 5
        ListNode* common = createNode(8);
        common->next = createNode(4);
        common->next->next = createNode(5);

        // List A: 4 -> 1 -> common
        ListNode* headA = createNode(4);
        headA->next = createNode(1);
        headA->next->next = common;

        // List B: 5 -> 6 -> 1 -> common
        ListNode* headB = createNode(5);
        headB->next = createNode(6);
        headB->next->next = createNode(1);
        headB->next->next->next = common;

        ListNode* intersection = getIntersectionNode(headA, headB);
        printf("Intersection node value: " + (intersection != NULL ? intersection->val : "NULL") + " (Expected: 8)\n");
        return 0;
}

/*
 * Time Complexity: O(M + N) where M and N are lengths of lists A and B.
 * Space Complexity: O(1) auxiliary pointers.
 */
